File: Hearthstone-Script/src/main/java/club/xiaojiawei/config/ThreadPoolConfig.java
package club.xiaojiawei.config;

import club.xiaojiawei.bean.LogThread;
import club.xiaojiawei.bean.WritableThread;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * 线程池配置
 * @author 肖嘉威
 * @date 2023/7/5 13:35
 */
@Configuration
public class ThreadPoolConfig {

    @Bean
    public ScheduledThreadPoolExecutor launchProgramThreadPool(){
        return new ScheduledThreadPoolExecutor(1, new ThreadFactory() {
            private final AtomicInteger num = new AtomicInteger(0);
            @Override
            public Thread newThread(Runnable r) {
                return new LogThread(r, "LaunchProgramPool Thread-" + num.getAndIncrement());
            }
        }, new ThreadPoolExecutor.AbortPolicy());
    }

    @Bean
    public ScheduledThreadPoolExecutor listenFileThreadPool(){
        return new ScheduledThreadPoolExecutor(4, new ThreadFactory() {
            private final AtomicInteger num = new AtomicInteger(0);
            @Override
            public Thread newThread(Runnable r) {
                return new WritableThread(r, "ListenFilePool Thread-" + num.getAndIncrement());
            }
        }, new ThreadPoolExecutor.AbortPolicy());
    }

    @Bean
    public ScheduledThreadPoolExecutor extraThreadPool(){
        return new ScheduledThreadPoolExecutor(6, new ThreadFactory() {
            private final AtomicInteger num = new AtomicInteger(0);
            @Override
            public Thread newThread(Runnable r) {
                return new LogThread(r, "ExtraPool Thread-" + num.getAndIncrement());
            }
        }, new ThreadPoolExecutor.AbortPolicy());
    }

    @Bean
    public ThreadPoolExecutor coreThreadPool(){
        return new ThreadPoolExecutor(2, 2, 1, TimeUnit.SECONDS, new ArrayBlockingQueue<>(1), new ThreadFactory() {
            private final AtomicInteger num = new AtomicInteger(0);
            @Override
            public Thread newThread(Runnable r) {
                return new LogThread(r, "CorePool Thread-" + num.getAndIncrement());
            }
        }, new ThreadPoolExecutor.AbortPolicy());
    }

}

File: Hearthstone-Script-Base/src/main/kotlin/club/xiaojiawei/config/ThreadPoolConfig.kt
package club.xiaojiawei.config

import club.xiaojiawei.bean.ReadableThread
import club.xiaojiawei.bean.WritableThread
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * @author 肖嘉威 xjw580@qq.com
 * @date 2024/9/18 16:47
 */
val LAUNCH_PROGRAM_THREAD_POOL: ScheduledThreadPoolExecutor by lazy {
    ScheduledThreadPoolExecutor(2, object : ThreadFactory {
        private val num = AtomicInteger(0)
        override fun newThread(r: Runnable): Thread {
            return WritableThread(r, "LaunchProgramPool Thread-" + num.getAndIncrement())
        }
    }, ThreadPoolExecutor.AbortPolicy())
}

val LISTEN_LOG_THREAD_POOL: ScheduledThreadPoolExecutor by lazy {
    ScheduledThreadPoolExecutor(4, object : ThreadFactory {
        private val num = AtomicInteger(0)
        override fun newThread(r: Runnable): Thread {
            return WritableThread(r, "ListenLogPool Thread-" + num.getAndIncrement())
        }
    }, ThreadPoolExecutor.AbortPolicy())
}

val EXTRA_THREAD_POOL: ScheduledThreadPoolExecutor by lazy {
    ScheduledThreadPoolExecutor(10, object : ThreadFactory {
        private val num = AtomicInteger(0)
        override fun newThread(r: Runnable): Thread {
            return ReadableThread(r, "ExtraPool Thread-" + num.getAndIncrement())
        }
    }, ThreadPoolExecutor.AbortPolicy())
}

val CORE_THREAD_POOL: ThreadPoolExecutor by lazy {
    ThreadPoolExecutor(2, 2, 1, TimeUnit.SECONDS, ArrayBlockingQueue(1), object : ThreadFactory {
        private val num = AtomicInteger(0)
        override fun newThread(r: Runnable): Thread {
            return WritableThread(r, "CorePool Thread-" + num.getAndIncrement())
        }
    }, ThreadPoolExecutor.AbortPolicy())
}

val CALC_THREAD_POOL: ThreadPoolExecutor by lazy {
    ThreadPoolExecutor(
        Runtime.getRuntime().availableProcessors(),
        Runtime.getRuntime().availableProcessors() * 2,
        120,
        TimeUnit.SECONDS,
        ArrayBlockingQueue(8),
        object : ThreadFactory {
            private val num = AtomicInteger(0)
            override fun newThread(r: Runnable): Thread {
                return ReadableThread(r, "CalcPool Thread-" + num.getAndIncrement())
            }
        },
        ThreadPoolExecutor.AbortPolicy()
    )
}

fun submitExtra(block: () -> Unit): Future<*> {
    return EXTRA_THREAD_POOL.submit(block)
}

val VIRTUAL_THREAD_POOL: ExecutorService =
    Executors.newThreadPerTaskExecutor(Thread.ofVirtual().name("VPool Thread-", 0).factory());